### ISSUE TITLE
Test Crossplane at scale

### ISSUE THREAD

**Issue by negz** [crossplane/crossplane#2568]
### What problem are you facing?

We've historically deferred optimising for performance when building Crossplane in favor of optimising for simplicity and maintainability. The idea being that we'd start optimising for performance when folks start hitting scaling limits. That time has come. I've been using the `performance` label to track issues that pertain to Crossplane performance, and we now have [a few](https://github.com/crossplane/crossplane/issues?q=is%3Aopen+is%3Aissue+label%3Aperformance). The general theme seems to be that when you get into the hundreds and thousands of XRs, Crossplane keeps working but takes longer and longer to reconcile.

We have several ideas about ways to improve performance, but before we start tackling those I'd like to get a rough benchmark we can use to determine what impact our performance tweaks have.

### How could Crossplane help solve your problem?

We could build a repeatable performance test suite. Off the top of my head, I'm thinking of something like a Crossplane XR that would:

1. Create a broadly representative Kubernetes cluster - something a Crossplane user might feasibly use in production.
2. Install Crossplane on said cluster.
3. Install a Configuration that includes an XR composed of say 3 managed resources.
4. Creates ~1,000 instances of those managed resources.

I'd then like to find some metrics to measure - e.g.:

* Average time for XRs to become ready.
* API server request rate.
* ???

The first step will be identifying some managed resources that are free to create and won't hit quota issues.

**Comment by negz** [comment]
I've put together a repository to make these tests somewhat reproducible at https://github.com/negz/crossplane-scale. An initial test of 100 claims (each with 4 managed resources) shows that it took about 18 minutes for all claims to become ready. By contrast a single claim becomes ready in about 30 seconds.

**Comment by negz** [comment]
I made some basic (but onerous, refactoring wise) tweaks to the core Crossplane controllers and the time taken for a burst of 100 new claims (each with four composed resources) to become ready dropped from ~18 minutes to ~8 minutes.

<img width="1374" alt="pretty-colours" src="https://user-images.githubusercontent.com/1049349/134293393-7e5bcbdb-84c8-44a0-b2d8-b00b79258185.png">

Specifically I:
* Allowed all controllers to process up to 10 reconciles concurrently. Previously some allowed 5, but most 1.
* Set all controllers over to use exponential backoff (gated by a global ~1rps rate limiter). Previously they used fix 30 second retries.

The managed resource controllers are already using exponential backoff (@hasheddan  did that a while back) but they’re all limited to one concurrent reconcile, so I suspect tweaking that will have a huge impact too.

There are some unexpected results on the graph though - we’re somehow hitting ~15 reconciles per second, when I’d expect it to be capped at ~3 reconciles per second (plus some bursts) given that we’re running three controller managers, each of which should be capped at ~1rps. I think technically the limit is on the rate at which reconciles are added to the work queue, so maybe there’s some head of line blocking going on? Not sure.

**Comment by negz** [comment]
> we’re somehow hitting ~15 reconciles per second, when I’d expect it to be capped at ~3 reconciles per second (plus some bursts) given that we’re running three controller managers, each of which should be capped at ~1rps.

I think there's potentially a couple of things happening here. Namely:

1. Not all reconciles are actually rate limited. There's four things that trigger a reconcile: a watch event, returning `RequeueAfter: t`, returning `Requeue: true`, or returning an error. Only the latter two cases are subject to rate limiting. That is, if you create 100 claims at once you'll get 100 claims queued at once (subject to max concurrency).
2. The rate limiter limits the rate at which things are queued to be reconciled, not the rate at which things are reconciled. The rate of queue processing is constrained by max concurrency. We use a `MaxOf` rate limiter that works by returning a time after which things may be added to our work queue, and in some cases the "per event" reconcile delay for a set of resources might all be (e.g.) T plus 30 seconds, so all of those resources will be queued for T plus 30 seconds regardless of what the global token bucket rate limiter says (because that rate limiter would never delay for more than 10 seconds).


I think this is either:

* Due to https://github.com/kubernetes/kubernetes/pull/24052#discussion_r60505934.

**Comment by negz** [comment]
I've been thinking of rate limiting as more of something that works like https://github.com/kubernetes-sigs/controller-runtime/issues/857.

**Comment by negz** [comment]
I've updated https://github.com/crossplane/crossplane/pull/2602 with an implementation that gets us 'true' global rate limiting - i.e. `--max-reconcile-rate=10` (the default) caps the controller manager at about 10rps. Combined with https://github.com/crossplane/provider-gcp/pull/380 my testing (still using https://github.com/negz/crossplane-scale) shows:

* A single claim takes < 30 seconds to become ready.
* A burst of 100 claims takes ~3.5 minutes to become ready.
* A burst of 1000 claims takes ~25 minutes to become ready.

This is a big improvement compared to what's in master at the time of writing, but there are likely further improvements. One interesting observation is that when I create 1,000 GCP bucket managed resources they all sit in the workqueue for ~10 minutes before it begins to actually burn through them. Note that it's likely they're all being reconciled and put back on the queue during this time - i.e. while we wait to confirm they exist. I wonder whether GCP was rate limiting us during this time, but couldn't immediately find evidence of that being the case.

<img width="1365" alt="buckets" src="https://user-images.githubusercontent.com/1049349/137986848-7901e1f0-76e9-4876-9c33-d85becb0925e.png">

**Comment by negz** [comment]
I'm going to close this out (but keep https://github.com/crossplane/crossplane/issues/2595), since I think the testing leading up to this is pretty much wrapped up.
